rust/otap-dataflow/crates/otap/src/experimental/azure_monitor_exporter/exporter.rs

```diff
@@ -154,30 +424,223 @@ impl Exporter<OtapPdata> for AzureMonitorExporter {
-                    ));
+            // Determine if we should accept new messages
+            let at_capacity = self.in_flight_exports.len() >= MAX_IN_FLIGHT_EXPORTS;
+
+            tokio::select! {
+                biased;
+
+                _ = tokio::time::sleep_until(next_token_refresh) => {
+                    let token = auth.get_token()
+                        .await
+                        .map_err(|e| Error::InternalError { message: format!("Failed to refresh token: {e}") })?;
+
+                    next_token_refresh = Self::get_next_token_refresh(token);
                 }
-                Message::PData(pdata) => {
-                    if let Err(e) = self.handle_pdata(pdata, &effect_handler).await {
-                        effect_handler
-                            .info(&format!(
-                                "[AzureMonitorExporter] Error processing data: {e}"
-                            ))
-                            .await;
+
+                completed = self.in_flight_exports.next_completion() => {
+                    if let Some(completed_export) = completed {
+                        self.finalize_export(&effect_handler, completed_export).await?;
+                    }
+                }
+
+                _ = tokio::time::sleep_until(next_periodic_export), if !at_capacity => {
+                    next_periodic_export = tokio::time::Instant::now() + tokio::time::Duration::from_secs(PERIODIC_EXPORT_INTERVAL);
+
+                    if self.last_batch_queued_at.elapsed() >= std::time::Duration::from_secs(PERIODIC_EXPORT_INTERVAL) && self.gzip_batcher.has_pending_data() {
+                        println!("[AzureMonitorExporter] Periodic export pending data");
+                        self.queue_current_batch(&effect_handler).await?;
+                    }
+                }
+
+                msg = msg_chan.recv(), if !at_capacity => {
+                    self.stats.message_received(self.in_flight_exports.len());
+
+                    match msg {
+                        Ok(Message::Control(NodeControlMsg::Shutdown { deadline, .. })) => {
+                            self.handle_shutdown(&effect_handler).await?;
+                            return Ok(TerminalState::new(
+                                deadline,
+                                std::iter::empty::<otap_df_telemetry::metrics::MetricSetSnapshot>(),
+                            ));
+                        }
+                        other => {
+                            self.handle_message(&effect_handler, other, &mut msg_id).await?;
+                        }
                     }
                 }
-                _ => {
-                    // Ignore other message types
+
+                _ = tokio::time::sleep_until(next_stats_print) => {
+                    next_stats_print = tokio::time::Instant::now() + tokio::time::Duration::from_secs(STATS_PRINT_INTERVAL);
```

Comment:
OK for experimental. We'll want to use the built-in metrics, and we'll want to make the continuous benchmark support this exporter as we've discussed.